By default ignore RRef leaks during shutdown

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30217 By default ignore RRef leaks during shutdown**

Before this commit, RRefContext throws an error if it detects any
RRef leak during shutdown. However, this requires applications to
make sure that is has freed all references to RRefs in application
code, which can be a bad debugging experience when for large
applications. Besides, this also relies on Python GC to free things
up in time, which might not always be true. After this commit,
RRefContext would ignore leaking RRefs during shutdown, as shutdown
is called when the application has finished training and no longer
care about local states. Hence, it should be OK to just ignore
those leaks and destroy OwnerRRefs. If application would like to
enforce no leaks, just set torch.distributed.rpc.api._ignore_rref_leak
to False.

closes #30105

Differential Revision: [D18632546](https://our.internmc.facebook.com/intern/diff/D18632546)

mrshenli committed:
By default ignore RRef leaks during shutdown

Before this commit, RRefContext throws an error if it detects any
RRef leak during shutdown. However, this requires applications to
make sure that is has freed all references to RRefs in application
code, which can be a bad debugging experience when for large
applications. Besides, this also relies on Python GC to free things
up in time, which might not always be true. After this commit,
RRefContext would ignore leaking RRefs during shutdown, as shutdown
is called when the application has finished training and no longer
care about local states. Hence, it should be OK to just ignore
those leaks and destroy OwnerRRefs. If application would like to
enforce no leaks, just set torch.distributed.rpc.api._ignore_rref_leak
to False.

[ghstack-poisoned]

zzzwen commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -7,13 +7,23 @@ namespace torch {
+    RRefContext::destroyed_ = true;
+  }
+  if (!ignoreRRefLeak) {
+    RRefContext::getInstance().checkRRefLeaks();
```
Q: How long do we want to keep this bypass available? If we are to keep this "ignore leak" bypass for a while, it might be better to print out the leak but not throw exception.

mrshenli replied:
It will be there for a while. Yes, let me add the print.

mrshenli committed:
Update on "By default ignore RRef leaks during shutdown"


Before this commit, RRefContext throws an error if it detects any
RRef leak during shutdown. However, this requires applications to
make sure that is has freed all references to RRefs in application
code, which can be a bad debugging experience when for large
applications. Besides, this also relies on Python GC to free things
up in time, which might not always be true. After this commit,
RRefContext would ignore leaking RRefs during shutdown, as shutdown
is called when the application has finished training and no longer
care about local states. Hence, it should be OK to just ignore
those leaks and destroy OwnerRRefs. If application would like to
enforce no leaks, just set torch.distributed.rpc.api._ignore_rref_leak
to False.

closes #30105

Differential Revision: [D18632546](https://our.internmc.facebook.com/intern/diff/D18632546)

[ghstack-poisoned]

zzzwen commented on test/rpc_test.py:
```diff
@@ -1091,6 +1091,51 @@ def test_call_method_on_rref(self):
+            api._ignore_rref_leak = True
+            rpc.wait_all_workers()
+        else:
+            with self.assertRaisesRegex(RuntimeError, "Leaking RRef"):
```
Nit, I think this assert should be part of `test_ignore_rref_leak`

mrshenli replied:
Could you please elaborate? Why `test_ignore_rref_leak` would expect a runtime error? Or do you mean sth like `self.assertWarnsRegex` after we add the warning?

zzzwen replied:
My bad, I mean move to `test_rref_leak`

pritamdamania87 commented on test/rpc_test.py:
```diff
@@ -1091,6 +1091,51 @@ def test_call_method_on_rref(self):
+        # This is for the below `dist.barrier`.
+        # For `RpcAgent` other than `ProcessGroupAgent`,
+        # no `_default_pg` is initialized.
+        if not dist.is_initialized():
```
I added a `_initialize_pg` helper method in `dist_autograd_test.py` for this. Maybe we can move that to dist_utils.py and use it in both tests?

mrshenli replied:
Will do dedup in a followup PR

pritamdamania87 commented on torch/csrc/distributed/rpc/rref.cpp:
```diff
@@ -136,16 +136,7 @@ UserRRef<T>::UserRRef(
-    fm->addCallback(
-        [](const Message& message) { RRefContext::handleException(message); });
-  }
+  RRefContext::getInstance().delUser(ownerId_, rrefId_, forkId_);
```
Can we catch and log exceptions here? Would be pretty risky to throw exceptions from the destructor.

mrshenli committed:
Update on "By default ignore RRef leaks during shutdown"


Before this commit, RRefContext throws an error if it detects any
RRef leak during shutdown. However, this requires applications to
make sure that is has freed all references to RRefs in application
code, which can be a bad debugging experience when for large
applications. Besides, this also relies on Python GC to free things
up in time, which might not always be true. After this commit,
RRefContext would ignore leaking RRefs during shutdown, as shutdown
is called when the application has finished training and no longer
care about local states. Hence, it should be OK to just ignore
those leaks and destroy OwnerRRefs. If application would like to
enforce no leaks, just set torch.distributed.rpc.api._ignore_rref_leak
to False.

closes #30105

Differential Revision: [D18632546](https://our.internmc.facebook.com/intern/diff/D18632546)

[ghstack-poisoned]

mrshenli committed:
Update on "By default ignore RRef leaks during shutdown"


Before this commit, RRefContext throws an error if it detects any
RRef leak during shutdown. However, this requires applications to
make sure that is has freed all references to RRefs in application
code, which can be a bad debugging experience when for large
applications. Besides, this also relies on Python GC to free things
up in time, which might not always be true. After this commit,
RRefContext would ignore leaking RRefs during shutdown, as shutdown
is called when the application has finished training and no longer
care about local states. Hence, it should be OK to just ignore
those leaks and destroy OwnerRRefs. If application would like to
enforce no leaks, just set torch.distributed.rpc.api._ignore_rref_leak
to False.

closes #30105

Differential Revision: [D18632546](https://our.internmc.facebook.com/intern/diff/D18632546)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/rref.cpp:
```diff
@@ -136,7 +136,19 @@ UserRRef<T>::UserRRef(
+  } catch(const std::exception& ex) {
+    LOG(ERROR) << "Error occurred when deleting UserRRef instance, "
+        << "RRefId = " << rrefId_ << ", ForkId = " << forkId_ << " : "
+        << ex.what() << std::endl;
```
LOG messages don't need std::endl.

pritamdamania87 commented on torch/csrc/distributed/rpc/rref.cpp:
```diff
@@ -136,7 +136,19 @@ UserRRef<T>::UserRRef(
+    LOG(ERROR) << "Error occurred when deleting UserRRef instance, "
+        << "RRefId = " << rrefId_ << ", ForkId = " << forkId_ << " : "
+        << (p ? p.__cxa_exception_type()->name() : "unknown error")
+        << std::endl;
```
same as above.

pritamdamania87 commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -55,7 +53,21 @@ void RRefContext::checkRRefLeaks() {
-    AT_ERROR(ss.str());
+
+    if (ignoreRRefLeak) {
+      LOG(WARNING) << "Detected RRef Leaks durinig shutdown. This usually "
```
nit: during.

pritamdamania87 commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -55,7 +53,21 @@ void RRefContext::checkRRefLeaks() {
+    if (ignoreRRefLeak) {
+      LOG(WARNING) << "Detected RRef Leaks durinig shutdown. This usually "
+          << "occurs when the application code still holds references to RRef "
+          << "instances when calling `shutdown()`. If the program has "
```
nit: backticks in LOG messages won't show up as intended on the terminal

pritamdamania87 commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -55,7 +53,21 @@ void RRefContext::checkRRefLeaks() {
+          << "after this, these leaks could result in memory leaks on RRef "
+          << "owners. Please make sure all RRefs are out of scope and Python "
+          << "GC has deleted them before calling `shutdown()`: \n"
+          << ss.str() << std::endl;
```
nit: no need for endl

mrshenli committed:
Update on "By default ignore RRef leaks during shutdown"


Before this commit, RRefContext throws an error if it detects any
RRef leak during shutdown. However, this requires applications to
make sure that is has freed all references to RRefs in application
code, which can be a bad debugging experience when for large
applications. Besides, this also relies on Python GC to free things
up in time, which might not always be true. After this commit,
RRefContext would ignore leaking RRefs during shutdown, as shutdown
is called when the application has finished training and no longer
care about local states. Hence, it should be OK to just ignore
those leaks and destroy OwnerRRefs. If application would like to
enforce no leaks, just set torch.distributed.rpc.api._ignore_rref_leak
to False.

closes #30105

Differential Revision: [D18632546](https://our.internmc.facebook.com/intern/diff/D18632546)

[ghstack-poisoned]

mrshenli committed:
Update on "By default ignore RRef leaks during shutdown"


Before this commit, RRefContext throws an error if it detects any
RRef leak during shutdown. However, this requires applications to
make sure that is has freed all references to RRefs in application
code, which can be a bad debugging experience when for large
applications. Besides, this also relies on Python GC to free things
up in time, which might not always be true. After this commit,
RRefContext would ignore leaking RRefs during shutdown, as shutdown
is called when the application has finished training and no longer
care about local states. Hence, it should be OK to just ignore
those leaks and destroy OwnerRRefs. If application would like to
enforce no leaks, just set torch.distributed.rpc.api._ignore_rref_leak
to False.

closes #30105

Differential Revision: [D18632546](https://our.internmc.facebook.com/intern/diff/D18632546)

[ghstack-poisoned]

mrshenli committed:
Update on "By default ignore RRef leaks during shutdown"


Before this commit, RRefContext throws an error if it detects any
RRef leak during shutdown. However, this requires applications to
make sure that is has freed all references to RRefs in application
code, which can be a bad debugging experience when for large
applications. Besides, this also relies on Python GC to free things
up in time, which might not always be true. After this commit,
RRefContext would ignore leaking RRefs during shutdown, as shutdown
is called when the application has finished training and no longer
care about local states. Hence, it should be OK to just ignore
those leaks and destroy OwnerRRefs. If application would like to
enforce no leaks, just set torch.distributed.rpc.api._ignore_rref_leak
to False.

closes #30105

Differential Revision: [D18632546](https://our.internmc.facebook.com/intern/diff/D18632546)

[ghstack-poisoned]

mrshenli commented:
@pytorchbot retest this please

mrshenli committed:
Update on "By default ignore RRef leaks during shutdown"


Before this commit, RRefContext throws an error if it detects any
RRef leak during shutdown. However, this requires applications to
make sure that is has freed all references to RRefs in application
code, which can be a bad debugging experience when for large
applications. Besides, this also relies on Python GC to free things
up in time, which might not always be true. After this commit,
RRefContext would ignore leaking RRefs during shutdown, as shutdown
is called when the application has finished training and no longer
care about local states. Hence, it should be OK to just ignore
those leaks and destroy OwnerRRefs. If application would like to
enforce no leaks, just set torch.distributed.rpc.api._ignore_rref_leak
to False.

closes #30105

Differential Revision: [D18632546](https://our.internmc.facebook.com/intern/diff/D18632546)

[ghstack-poisoned]

mrshenli committed:
Update on "By default ignore RRef leaks during shutdown"


Before this commit, RRefContext throws an error if it detects any
RRef leak during shutdown. However, this requires applications to
make sure that is has freed all references to RRefs in application
code, which can be a bad debugging experience when for large
applications. Besides, this also relies on Python GC to free things
up in time, which might not always be true. After this commit,
RRefContext would ignore leaking RRefs during shutdown, as shutdown
is called when the application has finished training and no longer
care about local states. Hence, it should be OK to just ignore
those leaks and destroy OwnerRRefs. If application would like to
enforce no leaks, just set torch.distributed.rpc.api._ignore_rref_leak
to False.

closes #30105

Differential Revision: [D18632546](https://our.internmc.facebook.com/intern/diff/D18632546)

[ghstack-poisoned]

mrshenli commented:
macos test failure is irrelevant:

```
Nov 25 09:44:18 Traceback (most recent call last):
Nov 25 09:44:18   File "test_cuda.py", line 21, in <module>
Nov 25 09:44:18     from test_torch import _TestTorchMixin
Nov 25 09:44:18   File "/Users/distiller/project/test/test_torch.py", line 54, in <module>
Nov 25 09:44:18     import librosa
Nov 25 09:44:18   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/librosa/__init__.py", line 12, in <module>
Nov 25 09:44:18     from . import core
Nov 25 09:44:18   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/librosa/core/__init__.py", line 124, in <module>
Nov 25 09:44:18     from .audio import *  # pylint: disable=wildcard-import
Nov 25 09:44:18   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/librosa/core/audio.py", line 10, in <module>
Nov 25 09:44:18     import soundfile as sf
Nov 25 09:44:18   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/soundfile.py", line 163, in <module>
Nov 25 09:44:18     _path, '_soundfile_data', _libname))
Nov 25 09:44:18 OSError: cannot load library '/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/_soundfile_data/libsndfile.dylib': dlopen(/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/_soundfile_data/libsndfile.dylib, 2): image not found
Nov 25 09:44:18 Traceback (most recent call last):
Nov 25 09:44:18   File "test/run_test.py", line 455, in <module>
Nov 25 09:44:18     main()
Nov 25 09:44:18   File "test/run_test.py", line 448, in main
Nov 25 09:44:18     raise RuntimeError(message)
Nov 25 09:44:18 RuntimeError: test_cuda failed!
```

mrshenli committed:
Update on "By default ignore RRef leaks during shutdown"


Before this commit, RRefContext throws an error if it detects any
RRef leak during shutdown. However, this requires applications to
make sure that is has freed all references to RRefs in application
code, which can be a bad debugging experience when for large
applications. Besides, this also relies on Python GC to free things
up in time, which might not always be true. After this commit,
RRefContext would ignore leaking RRefs during shutdown, as shutdown
is called when the application has finished training and no longer
care about local states. Hence, it should be OK to just ignore
those leaks and destroy OwnerRRefs. If application would like to
enforce no leaks, just set torch.distributed.rpc.api._ignore_rref_leak
to False.

closes #30105

Differential Revision: [D18632546](https://our.internmc.facebook.com/intern/diff/D18632546)

[ghstack-poisoned]

mrshenli committed:
Update on "By default ignore RRef leaks during shutdown"


Before this commit, RRefContext throws an error if it detects any
RRef leak during shutdown. However, this requires applications to
make sure that is has freed all references to RRefs in application
code, which can be a bad debugging experience when for large
applications. Besides, this also relies on Python GC to free things
up in time, which might not always be true. After this commit,
RRefContext would ignore leaking RRefs during shutdown, as shutdown
is called when the application has finished training and no longer
care about local states. Hence, it should be OK to just ignore
those leaks and destroy OwnerRRefs. If application would like to
enforce no leaks, just set torch.distributed.rpc.api._ignore_rref_leak
to False.

closes #30105

Differential Revision: [D18632546](https://our.internmc.facebook.com/intern/diff/D18632546)

[ghstack-poisoned]

mrshenli committed:
Update on "By default ignore RRef leaks during shutdown"


Before this commit, RRefContext throws an error if it detects any
RRef leak during shutdown. However, this requires applications to
make sure that is has freed all references to RRefs in application
code, which can be a bad debugging experience when for large
applications. Besides, this also relies on Python GC to free things
up in time, which might not always be true. After this commit,
RRefContext would ignore leaking RRefs during shutdown, as shutdown
is called when the application has finished training and no longer
care about local states. Hence, it should be OK to just ignore
those leaks and destroy OwnerRRefs. If application would like to
enforce no leaks, just set torch.distributed.rpc.api._ignore_rref_leak
to False.

closes #30105

Differential Revision: [D18632546](https://our.internmc.facebook.com/intern/diff/D18632546)

[ghstack-poisoned]
